python/cog/server/runner.py

```diff
@@ -363,6 +369,13 @@ async def async_predict_handling_errors() -> schema.PredictionResponse:
                 # since this is just before this coroutine exits
                 self._predictions.pop(request.id)
 
+                # all predictions in flight have been processed, we can exit
+                if self._shutting_down and not self._predictions:
+                    # note: this will do a bunch of work
+                    # and potentially slow down the coroutine completing
+                    # however, this should be okay
+                    self.terminate()
```

Comment:
I don't think I understand this. In `main` the workflow is:

- any thread can request a shutdown by setting the `shutdown_event`
- when that happens, `Server.stop` is called, which starts a shutdown of fastapi
- when fastapi shuts down it triggers a `shutdown` event, which calls `runner.shutdown()`
- the runner shutdown terminates the worker and stops the runner thread
- if this sequence fails to complete within 10 seconds, server shoots itself in the head by doing `os.kill(os.getpid(), signal.SIGKILL)`

If I understand this correctly, this is happening in a per-prediction coroutine. It doesn't make any sense to me at all to call `self.terminate` when `self._shutting_down` is already set.

I think instead what's missing here is a way for the `shutdown` method itself to wait for all its child coroutines to complete.